lib/mkDummySrc.nix

```diff
@@ -78,17 +78,25 @@ let
   # directory, we check if it happens to be an ancestor for an interesting file (i.e. is a prefix of
   # an interesting file). That way we are left with the smallest possible source needed for our
   # dummy derivation, and we bring any cache invalidation to a minimum. Whew!
-  mkBasePath = p: (toString p) + "/";
-  uncleanSrcBasePath = mkBasePath src;
 
-  uncleanFiles = findCargoFiles src;
+  # NB: if the `src` we were provided was filtered, make sure that we crawl the `origSrc`! Otherwise
+  # when we try to crawl the source Nix will evaluate the filter(s) fully resulting in a store path
+  # whose prefix won't match the paths we observe when we try to clean the source a bit further down
+  # (Nix optimizes multiple filters by running them all once against the original source).
+  # https://github.com/ipetkov/crane/issues/46
+  origSrc =
+    if src ? _isLibCleanSourceWith
+    then src.origSrc
+    else src;
+
+  uncleanSrcBasePath = (toString origSrc) + "/";
```

Comment:
I assume this is done to forcefully evaluate the source derivation? Wouldn't this already be evaluated in the call to `findCargoFiles`?

Reply:
I don't have a good enough intuitive understanding of how Nix treats paths and strings (and coercing between the two) to explain why this is needed... I did try taking it out and a bunch of tests failed so it's load bearing at some capacity :sweat_smile: